front/lib/resources/user_resource.ts

```diff
@@ -0,0 +1,263 @@
+import type {
+  ModelId,
+  Result,
+  UserProviderType,
+  UserType,
+} from "@dust-tt/types";
+import { Err, Ok } from "@dust-tt/types";
+import type { Attributes, ModelStatic, Transaction } from "sequelize";
+
+import type { Authenticator } from "@app/lib/auth";
+import { User } from "@app/lib/models/user";
+import { BaseResource } from "@app/lib/resources/base_resource";
+import { MembershipModel } from "@app/lib/resources/storage/models/membership";
+import type { ReadonlyAttributesType } from "@app/lib/resources/storage/types";
+
+// Attributes are marked as read-only to reflect the stateless nature of our Resource.
+// eslint-disable-next-line @typescript-eslint/no-empty-interface, @typescript-eslint/no-unsafe-declaration-merging
+export interface UserResource extends ReadonlyAttributesType<User> {}
+
+// eslint-disable-next-line @typescript-eslint/no-unsafe-declaration-merging
+export class UserResource extends BaseResource<User> {
+  static model: ModelStatic<User> = User;
+
+  constructor(model: ModelStatic<User>, blob: Attributes<User>) {
+    super(User, blob);
+  }
+
+  static async makeNew(
+    blob: Omit<
+      Attributes<User>,
+      | "id"
+      | "createdAt"
+      | "updatedAt"
+      | "isDustSuperUser"
+      | "providerId"
+      | "imageUrl"
+    > &
+      Partial<Pick<Attributes<User>, "providerId" | "imageUrl">>
+  ): Promise<UserResource> {
+    const user = await User.create(blob);
+    return new this(User, user.get());
+  }
+
+  static async listByModelIds(ids: ModelId[]): Promise<UserResource[]> {
+    const users = await User.findAll({
+      where: {
+        id: ids,
+      },
+    });
+
+    return users.map((user) => new UserResource(User, user.get()));
+  }
+
+  static async listByUsername(username: string): Promise<UserResource[]> {
+    const users = await User.findAll({
+      where: {
+        username,
+      },
+    });
+
+    return users.map((user) => new UserResource(User, user.get()));
+  }
+
+  static async listByEmail(email: string): Promise<UserResource[]> {
+    const users = await User.findAll({
+      where: {
+        email,
+      },
+    });
+
+    return users.map((user) => new UserResource(User, user.get()));
+  }
+
+  static async fetchNonNullableByModelId(
```

Comment:
I'm not a big fan of this approach since it introduces more entry points and variations for the fetching method.